lib/ace/ext/papeeria/spellchecker.coffee

```diff
@@ -1,39 +1,70 @@
 define( ->
-  # This will be removed and replaced with real data from server.
-  # BTW, this is kinda template for data to send from server.
-  TEST_JSON_TYPOS = {
-    "blablabla": [
-      "blah-blah-blah"
-      "blabber"
-      "blabbed"
-      "salable"
-    ]
-  }
 
+  class Spellchecker
+    constructor: (@editor) ->
+      @_init(null)
+      @typosUrl = null          # url used to fetch typos
+      @suggestionsUrl = null    # url used to fetch corrections for a word
+      @typosHash = null         # hash used to check whether the typos list has been changed
+
+    _init: (language) =>
+      @language = language      # language code, e.g. `en_US`
+      @typos = {}               # set of typos
+
+    _fetchTypos: =>
+      $.getJSON(@typosUrl, null, (typosArray) =>
+        tmp = {}
+        for typo in typosArray
+          tmp[typo] = true
+        @typos = tmp
+        @editor.getSession()._emit("updateSpellcheckingTypos", {typos: typosArray})
+      )
+
+    # Update spellchecking settings
+    # @param {Object} settings: object with the following keys:
+    #        @param {String}           alphabet: language's alphabet, used for tokenizing
+    #        @param {Boolean}          enabled: whether spellchecking is enabled
+    #        @param {String}           language: language code, e.g. `en_US`
+    #        @param {Function}         onSettingsUpdateSuccess: success callback
+    #        @param {Function(String)} onSettingsUpdateError: error callback, takes error description
+    onSettingsUpdated: (settings) =>
+      @_init(settings.language)
+      @editor.getSession()._emit("changeSpellingCheckSettings", settings)
 
-  # Class by now provides words checking (used for highlighting)
-  # and also provides corrections list for a given word.
-  class SpellChecker
-    # By now it's just a stub, but I guess this function will take data from
-    # server somehow in future.
-    # @return {Object} JSON-list of incorrect words.
-    getJson: ->
-      return TEST_JSON_TYPOS
+    # Update spellchecking session
+    # @param {Object} session: object with the following keys:
+    #        @param {String} typosHash: hash used to check whether the typos list has been changed
+    #        @param {String} typosUrl: url used to fetch typos
+    #        @param {String} suggestionsUrl: url used to fetch corrections for a word
+    onSessionUpdated: (session) =>
+      @typosUrl = session.typosUrl
+      @suggestionsUrl = session.suggestionsUrl
+      if @typosHash != session.typosHash
+        @typosHash = session.typosHash
+        @_fetchTypos()
```

Comment:
I suggest swapping these calls and setting hash only on successful retrieval of typos